fv3core/stencils/mapn_tracer.py

```diff
@@ -53,64 +52,55 @@ def compute(
     qs = utils.make_storage_from_shape(
         pe1.shape, origin=(0, 0, 0), cache_key="mapn_tracer_qs"
     )
-    (
-        dp1,
-        q4_1,
-        q4_2,
-        q4_3,
-        q4_4,
-        origin,
-        domain,
-        i_extent,
-        j_extent,
-    ) = map_single.setup_data(tracers[utils.tracer_variables[0]], pe1, i1, i2, j1, j2)
 
-    # transliterated fortran 3d or 2d validate, not bit-for bit
+    map_single = utils.cached_stencil_class(MapSingle)(
+        spec.namelist, cache_key="mapntracer-single"
+    )
     tracer_list = [tracers[q] for q in utils.tracer_variables[0:nq]]
+    map_single.setup_data(tracer_list[0], pe1, i1, i2, j1, j2)
+
     for tracer in tracer_list:
         set_components(
             tracer,
-            q4_1,
-            q4_2,
-            q4_3,
-            q4_4,
+            map_single.q4_1,
+            map_single.q4_2,
+            map_single.q4_3,
+            map_single.q4_4,
             origin=(spec.grid.is_, spec.grid.js, 0),
             domain=domain_compute,
         )
 
         q4_1, q4_2, q4_3, q4_4 = remapping_calculation(
             qs,
-            q4_1,
-            q4_2,
-            q4_3,
-            q4_4,
-            dp1,
+            map_single.q4_1,
+            map_single.q4_2,
+            map_single.q4_3,
+            map_single.q4_4,
+            map_single.dp1,
             i1,
             i2,
             j1,
             j2,
             q_min,
         )
-        map_single.do_lagrangian_contributions(
+        map_single.lagrangian_contributions(
             tracer,
             pe1,
             pe2,
             q4_1,
             q4_2,
             q4_3,
             q4_4,
-            dp1,
+            map_single.dp1,
             i1,
             i2,
             j1,
             j2,
-            kord,
-            origin,
-            domain,
-            version,
+            map_single.origin,
+            map_single.domain,
         )
```

Comment:
This is maybe more of a comment for the mapn_tracer oopification, but I think we could also get away with replacing the stuff above here in the for-loop with a call to MapSingle.

Again, completely fine for that to be a subsequent refactor!

Reply:
Moved the origin and domain into the stencil.